### ISSUE TITLE
libiconv: windres: Command not found

### ISSUE THREAD

**Issue by LinusChen-yf** [conan-io/conan-center-index#26445]
### What is your question?

An error occurs when cross-platform compiling a Windows package on a Linux system.
The package is `libiconv/1.17`
```
libtool: compile:  x86_64-w64-mingw32-gcc -I. -I/home/crash/.conan2/p/b/libic94830bd8c5ce2/b/src/lib -I../include -I/home/crash/.conan2/p/b/libic94830bd8c5ce2/b/src/lib/../include -I.. -I/home/crash/.conan2/p/b/libic94830bd8c5ce2/b/src/lib/.. -DNDEBUG -m64 -O3 -DLIBDIR=\"//lib\" -DBUILDING_LIBICONV -DBUILDING_LIBCHARSET -DBUILDING_DLL -DENABLE_RELOCATABLE=1 -DIN_LIBRARY -DINSTALLDIR=\"//lib\" -DNO_XMALLOC -Dset_relocation_prefix=libiconv_set_relocation_prefix -Drelocate=libiconv_relocate -Drelocate2=libiconv_relocate2 -DHAVE_CONFIG_H -c /home/crash/.conan2/p/b/libic94830bd8c5ce2/b/src/lib/relocatable.c -o relocatable.o
libtool: compile:  windres --target=pe-x86-64 -DPACKAGE_VERSION_STRING=\\\"1.17\\\" -DPACKAGE_VERSION_MAJOR=1 -DPACKAGE_VERSION_MINOR=17 -DPACKAGE_VERSION_SUBMINOR=0 -i /home/crash/.conan2/p/b/libic94830bd8c5ce2/b/src/lib/../windows/libiconv.rc --output-format=coff -o libiconv.res.o
../libtool: line 1932: windres: command not found
make[1]: *** [Makefile:99: libiconv.res.lo] Error 1
make[1]: *** Waiting for unfinished jobs....
```

I have modified the profile file according to another issue, but an error has been reported. conan-io/conan#14635
But I compiled the package `zlib/1.3.1` of this issue successfully.

My host profile:
```
[settings]
os=Windows
arch=x86_64
compiler=gcc
compiler.cppstd=gnu14
compiler.libcxx=libstdc++11
compiler.version=9
build_type=Release

[buildenv]
CONAN_CMAKE_FIND_ROOT_PATH=/usr/x86_64-w64-mingw32
CHOST=x86_64-w64-mingw32
AR=x86_64-w64-mingw32-ar
AS=x86_64-w64-mingw32-as
RANLIB=x86_64-w64-mingw32-ranlib
CC=x86_64-w64-mingw32-gcc
CXX=x86_64-w64-mingw32-g++
STRIP=x86_64-w64-mingw32-strip
RC=x86_64-w64-mingw32-windres
```

My build profile:
```
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=gnu17
compiler.libcxx=libstdc++11
compiler.version=14
os=Linux
```
And my x86_64-w64-mingw32-windres program exists.
```
> which x86_64-w64-mingw32-windres
> /usr/bin/x86_64-w64-mingw32-windres
```

I have searched the internet for a long time but couldn't find a solution. Thank you very much for your help.

### Have you read the CONTRIBUTING guide?

- [x] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Transferred the ticket to ``conan-center-index`` repo, to try to address the issue in the ``libiconv`` recipe

**Comment by jcar87** [comment]
@amachado-pie , @LinusChen-yf 

INdeed it's possible that the recipe is hardcoding an executable name that it expects to find in path (although I think this may be for good reason)

- Are you able to build the recipe if you comment out the lines in the recipe that define `"WINDRES": "windres --target=pe-x86-64",`?

How is the mingw toolchain installed on this Linux system? Sometimes there's two sets of executables, the ones "prefixed", e.g. `x86_64-w64-mingw32-windres` , and the ones without a prefix, e.g `windres` - one is typically a symlink to the other, and one may be in a `libexec/bin` folder or similar

If there's a chance you have the "unprefixed" executables in a bild folder, it's worth trying to add that to the path (in the `[buildenv]`) section - otherwise I think we're looking at fixing this in the recipe

if you could confirm how you've installed the mingw toolchain (e.g. is this an ubuntu package?) - we can reproduce this to work on a fix

**Comment by amachado-pie** [comment]
Hi @jcar87  thanks for look into this.

Inded, we have installed, in addition to gcc / g++, the following Ubuntu packages for cross compiling to Windows:
https://launchpad.net/ubuntu/+source/gcc-mingw-w64
```
gcc-mingw-w64-i686: GNU C compiler for MinGW-w64 targeting Win32
g++-mingw-w64-i686: GNU C++ compiler for MinGW-w64 targeting Win32
gcc-mingw-w64-x86-64: GNU C compiler for MinGW-w64 targeting Win64
g++-mingw-w64-x86-64: GNU C++ compiler for MinGW-w64 targeting Win64
```
The binaries of these packages are being installed at `/bin`
```
which i686-w64-mingw32-gcc x86_64-w64-mingw32-gcc gcc 
/usr/bin/i686-w64-mingw32-gcc
/usr/bin/x86_64-w64-mingw32-gcc
/usr/bin/gcc
```

So we are just defining in the conan profile and cmake toolchain file, for each target, the toolchain binaries like this:

`conan_profiles/windows-x86`:
```
[settings]
os=Windows
arch=x86
build_type=Release
compiler=gcc
compiler.cppstd=11
compiler.libcxx=libstdc++11
compiler.version=6.3

[conf]
tools.cmake.cmaketoolchain:user_toolchain=["{{profile_dir}}/../cmake/toolchains/windows-x86.toolchain.cmake"]
tools.build:compiler_executables={"c":"i686-w64-mingw32-gcc","cpp":"i686-w64-mingw32-g++","rc":"i686-w64-mingw32-windres"}

[buildenv]
CFLAGS=-m32 -march=pentium-mmx
CXXFLAGS=-m32 -march=pentium-mmx
LDFLAGS=-m32

AS=i686-w64-mingw32-as
LD=i686-w64-mingw32-gcc
CC=i686-w64-mingw32-gcc
CXX=i686-w64-mingw32-g++
AR=i686-w64-mingw32-ar
NM=i686-w64-mingw32-nm
STRIP=i686-w64-mingw32-strip
OBJCOPY=i686-w64-mingw32-objcopy
OBJDUMP=i686-w64-mingw32-objdump
RANLIB=i686-w64-mingw32-ranlib
RC=i686-w64-mingw32-windres
```

`cmake/toolchains/windows-x86.toolchain.cmake`:
```
set(CMAKE_SYSTEM_NAME Windows)
set(CMAKE_SYSTEM_PROCESSOR x86)

set(CMAKE_C_COMPILER i686-w64-mingw32-gcc)
set(CMAKE_CXX_COMPILER i686-w64-mingw32-g++)
set(CMAKE_ASM_COMPILER i686-w64-mingw32-as)
set(CMAKE_AR i686-w64-mingw32-ar)
set(CMAKE_NM i686-w64-mingw32-nm)
set(CMAKE_RANLIB i686-w64-mingw32-ranlib)
set(CMAKE_STRIP i686-w64-mingw32-strip)
set(CMAKE_OBJCOPY i686-w64-mingw32-objcopy)
set(CMAKE_OBJDUMP i686-w64-mingw32-objdump)
set(CMAKE_RC_COMPILER i686-w64-mingw32-windres)
```

Are we doing something wrong and should we define the toolchains using path with the symbolic links?

What I don't understand is if we are defining an RC variable with the path/name of the binary to be used to compile the resources in the conan profile and in the cmake toolchain file, the recipe is ignoring it and expects the existence of the binary with a fixed name.

Thanks!

**Comment by jcar87** [comment]
Thanks @amachado-pie  - this is helpful for us to reproduce.

> resources in the conan profile and in the cmake toolchain file, the recipe is ignoring it and expects the existence of the binary with a fixed name.

Bear in mind that defining `[tools.cmake.cmaketoolchain:user_toolchain` only has an effect on recipes that are built with CMake - the one reported in this issue is built with autotools, so it has no effect in this particular case.

`RC=i686-w64-mingw32-windres` as a buildenv should work, however this is contingent on the build system of the particular recipe obeying this setting like that - most standard Autotools will, but some that deviate (e.g. ffmpeg) might not. In this case, I suspect you may be correct and that https://github.com/conan-io/conan-center-index/pull/22665 is deviating from convention

**Comment by amachado-pie** [comment]
@jcar87 I don't know if that #22665 was an attempt to handle an issue when no RC defined or to set the `--target target` to specify the BFD format to use according arch 32bit/64bit.
May be an approach like this can resolve both issues (no tested yet):

```diff
--- a/recipes/libiconv/all/conanfile.py
+++ b/recipes/libiconv/all/conanfile.py
@@ -88,17 +88,20 @@ class LibiconvConan(ConanFile):
 
         tc = AutotoolsToolchain(self)
         if self.settings.os == "Windows" and self.settings.compiler == "gcc":
+            env = tc.environment();
+            rc = env.get("RC", "windres")
+
             if self.settings.arch == "x86":
                 tc.update_configure_args({
                     "--host": "i686-w64-mingw32",
-                    "RC": "windres --target=pe-i386",
-                    "WINDRES": "windres --target=pe-i386",
+                    "RC": f"{rc} --target=pe-i386",
+                    "WINDRES": f"{rc} --target=pe-i386",
                 })
             elif self.settings.arch == "x86_64":
                 tc.update_configure_args({
                     "--host": "x86_64-w64-mingw32",
-                    "RC": "windres --target=pe-x86-64",
-                    "WINDRES": "windres --target=pe-x86-64",
+                    "RC": f"{rc} --target=pe-x86-64",
+                    "WINDRES": f"{rc} --target=pe-x86-64",
                 })
```

**Comment by jcar87** [comment]
> [@jcar87](https://github.com/jcar87) I don't know if that [#22665](https://github.com/conan-io/conan-center-index/pull/22665) was an attempt to handle an issue when no RC defined or to set the `--target target` to specify the BFD format to use according arch 32bit/64bit. 

I suspect it's the latter - the `windres` executable itself is located by autotools, and it tries to locate it with different filenames until it finds it. I suspect the changes in the recipe are just a way to add those flags- which could possibly be done via `RCFLAGS` instead. However, it's still not convinced it is exactly needed:

The [documentation](https://sourceware.org/binutils/docs/binutils/windres.html) for `windres` says:

> --target target
> Specify the BFD format to use for a COFF file as input or output. This is a BFD target name; you can use the --help option to see a list of supported targets. Normally windres will use the default format, which is the **first one listed by the --help option**. [Target Selection](https://sourceware.org/binutils/docs/binutils/Target-Selection.html).

On Ubuntu:
```
/usr/bin/x86_64-w64-mingw32-windres: supported targets: pe-x86-64 ...
```

On msys2 - with `mingw-w64-x86_64-toolchain`
```
C:\msys64\mingw64\bin\windres.exe: supported targets: pe-x86-64 ...
```

msys2 with  the `mingw-w64-i686-toolchain`:
```
C:\msys64\mingw32\bin\windres.exe: supported targets: pe-i386 
```

So if the right windres is located, it should already be doing the right thing. If this fix was needed at all, chances are it's a misconfigured Conan profile, rather than a recipe issue
